Allow declaring exclusive resources for child nodes (#4151)
junit-team/junit5: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc

include::{testDir}/example/sharedresources/DynamicSharedResourcesDemo.java[tags=user_guide]
----

Also, "static" shared resources can be declared for _direct_ child nodes via the `target`
attribute in the `{ResourceLock}` annotation, the attribute accepts a value from
the `{ResourceLockTarget}` enum.

Using the `target = CHILDREN` in a class-level `{ResourceLock}` annotation
has the same semantics as adding an annotation with the same `value` and `mode`

> to each test method declared in this class.

Reviewer: ```suggestion
to each test method and nested test class declared in this class.
```

> It may improve parallelization when a test class declares a `READ` lock,

Reviewer: ```suggestion
This may improve parallelization when a test class declares a `READ` lock,
```

but only a few methods hold a `READ_WRITE` lock.

Tests in the following example would run in the `SAME_THREAD` if the `{ResourceLock}`

> doesn't have the `target = CHILDREN`. This is because the test class declares a `READ`

Reviewer: ```suggestion
didn't have `target = CHILDREN`. This is because the test class declares a `READ`
```

shared resource, but one test method holds a `READ_WRITE` lock,

> which forces the `SAME_THREAD` execution mode for all the test methods.

Reviewer: ```suggestion
which would force the `SAME_THREAD` execution mode for all the test methods.
```
